### DISCUSSION TITLE
Add a optional download button on Code Blocks

### DISCUSSION THREAD

**Discussion by jy95** [facebook/docusaurus#8457] [Ideas]
Hello,

For [code blocks](https://docusaurus.io/docs/markdown-features/code-blocks#syntax-highlighting), we already have a [copy button](https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-theme-classic/src/theme/CodeBlock/CopyButton/index.tsx) so why not a button to download the code block into a file ?

For example : 

```jsx
```jsx title="/src/components/HelloCodeTitle.js" canDownload=true
function HelloCodeTitle(props) {
  return <h1>Hello, {props.name}</h1>;
}
``` 

When clicking on the download button, you get a file called `HelloCodeTitle.js` ( or `src_components_HelloCodeTitle.js` to keep it simple)

As showncast in my example, that behavior should be disabled by default, as not all code blocks are worthy to be download IHMO
(One open question is how to provide the [MIME](https://developer.mozilla.org/en-US/docs/Glossary/MIME_type) type to [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) or having by default `application/text` in case nothing was provided )

What do you think ?

Thanks in advance

**Comment by slorber** [comment]
🤷‍♂️ never seen that feature before and I don't really find it valuable. Do you have example doc sites doing this?

It's quite niche and you can implement it on your own site with swizzle. If it's not possible or complex (too much to swizzle) than we should focus on making it easier for you to add this button, but I don't think we want it to be a core feature unless we have at least a few users requesting it, and proof that it's a good UX.

**Reply by jy95** [reply]
IHMO, I think it isn't niche and could be interesting to be added.
I don't have specific examples in mind but I can see situations showing its value : 
- https://mui.com/ : you want to download several variant of a Component & play with it locally
- You have multiple files to copy and it becomes annoying to copy them into files 
- ...
(It could be useful for Meta packages such as Jest that put bunch of code blocks - https://jestjs.io/docs/setup-teardown )

I don't think it is safe to swizzle the whole CodeBlock component just to add a button (in the [docs](https://docusaurus.io/docs/markdown-features/code-blocks#custom-magic-comments) usually you swizzle only a specific component such as Line or even CopyButton if you want for example to replace the icon ^^). It would be nice to have a `Buttons` components

```bash 
npm run swizzle @docusaurus/theme-classic CodeBlock/Buttons
``` 

Then in swizzle : 

```tsx
import CopyButton from '@theme/CodeBlock/CopyButton';

function MyComponent() {

   return (
        <Buttons>
                {/* Original Copy Button from Docusaurus */}
                <CopyButton />
                {/* TODO Extra buttons */}
                <DownloadButton >
        </Buttons>
  )
}
```

**Reply by slorber** [reply]
> IHMO, I think it isn't niche and could be interesting to be added.

If we have 5+ people asking for this we can consider it, once we have a proper understanding of the use-case, a design proposal and the bandwidth to implement it.

There's a canny feature request board where people can vote so it's a better place to collect interest.

Jest website maintainers (Simen and I) or other Meta sites haven't expressed interested in having this feature so far

> It would be nice to have a Buttons components

That's much more likely to happen in the short term, I'm in favor of enabling users to implement their opinionated stuff, over providing everything out of the box 👍

**Reply by jy95** [reply]
Where is this canny feature request board ?

As I said above, I acknowledge that all codes blocks aren't worthy to be downloaded into files, thus the `canDownload` field.

In my mind, stuff like : 
- examples configuration files (yaml, json, ...)
- queries ([graphql](https://graphql.org/code/#javascript), sql, ...)
- small files output (csv, ...)
- scripts (when introducing the notion of plugins, ...) 
- ...

that are supported in [prism](https://prismjs.com/#supported-languages) are scenarios where such feature find its interests ...

> That's much more likely to happen in the short term, I'm in favor of enabling users to implement their opinionated stuff, over providing everything out of the box

Open closed principle is welcome : some people might have funzy ideas such as a button to have a picture of the code with [Carbon app](https://carbon.now.sh/)

**Reply by slorber** [reply]
> Where is this canny feature request board ?

Feature Requests in the sidebar: https://docusaurus.io/community/support

**Reply by jy95** [reply]
If looking for an example (in the future), I saw this on website https://bundlejs.com/ 
![image](https://user-images.githubusercontent.com/9306961/212444668-483e8b01-d6cc-4bab-9b17-d39e397ef7ac.png)

**Reply by shanekik** [reply]
This is also my use case. I need to add a button to open up that piece of code in an external code playground.

**Comment by ross-spencer** [comment]
Added a +1, seems like a valuable addition.

**Reply by jy95** [reply]
Thanks, just an extra vote on https://docusaurus.io/feature-requests/p/add-a-optional-download-button-on-code-blocks and we are good to discuss it again
